File: phase-3-chatbot/backend/src/schemas/task.py
"""
Pydantic schemas for Task API requests and responses.

These schemas define the structure and validation rules for API data,
separate from the database models to provide clear API contracts.
"""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from src.models.task import RecurringType, TaskPriority


class TaskCreate(BaseModel):
    """Schema for creating a new task."""

    title: str = Field(..., min_length=1, max_length=200, description="Task title")
    description: str | None = Field(None, max_length=1000, description="Task description")
    priority: TaskPriority = Field(default=TaskPriority.medium, description="Task priority level")
    due_date: datetime | None = Field(None, description="Task deadline")
    recurring: RecurringType = Field(default=RecurringType.none, description="Recurring schedule type")
    recurring_end_date: datetime | None = Field(None, description="End date for recurring schedule")
    tags: list[str] = Field(default_factory=list, description="Task tags/categories")


class TaskUpdate(BaseModel):
    """Schema for full task update (PUT request). All fields required."""

    title: str = Field(..., min_length=1, max_length=200, description="Task title")
    description: str | None = Field(None, max_length=1000, description="Task description")
    complete: bool = Field(..., description="Task completion status")
    priority: TaskPriority = Field(..., description="Task priority level")
    due_date: datetime | None = Field(None, description="Task deadline")
    recurring: RecurringType = Field(default=RecurringType.none, description="Recurring schedule type")
    recurring_end_date: datetime | None = Field(None, description="End date for recurring schedule")
    tags: list[str] = Field(default_factory=list, description="Task tags/categories")


class TaskPatch(BaseModel):
    """Schema for partial task update (PATCH request). All fields optional."""

    title: str | None = Field(None, min_length=1, max_length=200, description="Task title")
    description: str | None = Field(None, max_length=1000, description="Task description")
    complete: bool | None = Field(None, description="Task completion status")
    priority: TaskPriority | None = Field(None, description="Task priority level")
    due_date: datetime | None = Field(None, description="Task deadline")
    recurring: RecurringType | None = Field(None, description="Recurring schedule type")
    recurring_end_date: datetime | None = Field(None, description="End date for recurring schedule")
    tags: list[str] | None = Field(None, description="Task tags/categories")


class TaskResponse(BaseModel):
    """Schema for task response (all endpoints return this)."""

    id: int
    user_id: str
    title: str
    description: str | None
    complete: bool
    priority: TaskPriority
    due_date: datetime | None
    recurring: RecurringType
    recurring_end_date: datetime | None
    tags: list[str]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TaskListResponse(BaseModel):
    """Schema for task list response."""

    tasks: list[TaskResponse]
    count: int


class TaskQueryParams(BaseModel):
    """Query parameters for searching, filtering, and sorting tasks."""

    search: str | None = Field(None, description="Keyword search in title/description")
    status: str | None = Field(None, description="Filter by status: all/pending/completed")
    priority: TaskPriority | None = Field(None, description="Filter by priority level")
    tags: list[str] | None = Field(None, description="Filter by tags")
    sort_by: str = Field(default="created_at", description="Sort field: created_at/due_date/priority/title")
    sort_order: str = Field(default="desc", description="Sort order: asc/desc")

File: phase-3-chatbot/frontend/types/task.ts
// Task types matching backend SQLModel schema

export type TaskPriority = 'low' | 'medium' | 'high';

export type RecurringType = 'none' | 'daily' | 'weekly' | 'monthly' | 'yearly';

export interface Task {
  id: number;                    // Unique task identifier (auto-increment from backend)
  user_id: string;               // Owner's user ID (foreign key to User.id)
  title: string;                 // Task title (required)
  description: string | null;    // Task description (optional)
  complete: boolean;             // Completion status
  priority: TaskPriority;        // Task priority (low/medium/high)
  due_date: string | null;       // ISO 8601 timestamp of deadline (optional)
  recurring: RecurringType;      // Recurring schedule type
  recurring_end_date: string | null; // ISO 8601 timestamp of recurring end (optional)
  tags: string[];                // Task tags/categories
  created_at: string;            // ISO 8601 timestamp of creation
  updated_at: string;            // ISO 8601 timestamp of last update
}

// Payload for creating a new task
export interface TaskCreate {
  title: string;                 // Required task title
  description?: string;          // Optional task description
  priority?: TaskPriority;       // Optional task priority (defaults to 'medium' on backend)
  due_date?: string;             // Optional due date (ISO 8601)
  recurring?: RecurringType;     // Optional recurring type (defaults to 'none')
  recurring_end_date?: string;   // Optional recurring end date
  tags?: string[];               // Optional tags
}

// Payload for full task update (PUT)
export interface TaskUpdate {
  title: string;                 // Updated title
  description: string | null;    // Updated description
  complete: boolean;             // Updated completion status
  priority: TaskPriority;        // Updated priority
  due_date?: string | null;      // Updated due date
  recurring?: RecurringType;     // Updated recurring type
  recurring_end_date?: string | null; // Updated recurring end date
  tags?: string[];               // Updated tags
}

// Payload for partial task update (PATCH)
export interface TaskPatch {
  title?: string;                // Partial update: title only
  description?: string | null;   // Partial update: description only
  complete?: boolean;            // Partial update: completion status only
  priority?: TaskPriority;       // Partial update: priority only
  due_date?: string | null;      // Partial update: due date
  recurring?: RecurringType;     // Partial update: recurring type
  recurring_end_date?: string | null; // Partial update: recurring end date
  tags?: string[];               // Partial update: tags
}

// Response from backend when fetching tasks
export interface TaskListResponse {
  tasks: Task[];                 // Array of task objects
  count: number;                 // Total number of tasks
}

// Single task response (same as Task)
export interface TaskResponse extends Task {}

// Query parameters for search/filter/sort
export interface TaskQueryParams {
  search?: string;
  status?: 'all' | 'pending' | 'completed';
  priority?: TaskPriority;
  tags?: string[];
  sort_by?: 'created_at' | 'due_date' | 'priority' | 'title';
  sort_order?: 'asc' | 'desc';
}
